noir-projects/aztec-nr/aztec/src/state_vars/public_immutable.nr

```diff
@@ -22,12 +25,53 @@ impl<T> PublicImmutable<T> {
         PublicImmutable { context, storage_slot }
     }
     // docs:end:public_immutable_struct_new
+}
+
+impl <T> PublicImmutable<T, &mut PublicContext> {
+    // docs:start:public_immutable_struct_write
+    pub fn initialize<T_SERIALIZED_LEN>(self, value: T) where T: Serialize<T_SERIALIZED_LEN> {
+        // TODO(#4738): Uncomment the following assert
+        // assert(
+        //     self.context.public.unwrap_unchecked().is_deployment(), "PublicImmutable can only be initialized during contract deployment"
+        // );
+
+        // We check that the struct is not yet initialized by checking if the initialization slot is 0
+        let initialization_slot = INITIALIZATION_SLOT_SEPARATOR + self.storage_slot;
+        let fields_read: [Field; 1] = storage_read(initialization_slot);
+        assert(fields_read[0] == 0, "PublicImmutable already initialized");
+
+        // We populate the initialization slot with a non-zero value to indicate that the struct is initialized
+        storage_write(initialization_slot, [0xdead]);
+
+        let fields_write = T::serialize(value);
+        storage_write(self.storage_slot, fields_write);
+    }
+    // docs:end:public_immutable_struct_write
+
+    // Note that we don't access the context, but we do call oracles that are only available in public
+    // docs:start:public_immutable_struct_read
+    pub fn read<T_SERIALIZED_LEN>(self) -> T where T: Deserialize<T_SERIALIZED_LEN> {
+        let fields = storage_read(self.storage_slot);
+        T::deserialize(fields)
+    }
+    // docs:end:public_immutable_struct_read
+}
+
+impl<T> PublicImmutable<T, ()> {
+    pub fn read<T_SERIALIZED_LEN>(self) -> T where T: Deserialize<T_SERIALIZED_LEN> {
+        // Note that this is the exact same implementation as for public execution, though it might change in the future
+        // since unconstrained execution might not rely on the same oracles as used for public execution (which 
+        // transpile to AVM opcodes).
+        let fields = storage_read(self.storage_slot);
+        T::deserialize(fields)
+    }
+}
 
+// TODO (https://github.com/AztecProtocol/aztec-packages/issues/5818): remove this impl and leave the PublicContext impl
+// once AvmContext becomes PublicContext.
+impl <T> PublicImmutable<T, &mut AvmContext> {
```

Comment:
This is simply a copy of the `PublicContext` impl. I added a note in https://github.com/AztecProtocol/aztec-packages/issues/5818 so that we don't forget to remove it once the transition is complete.